python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -710,6 +711,30 @@ def do_evaluate(
                 flags=plc.strings.regex_flags.RegexFlags.DEFAULT,
             )
             return Column(plc.strings.contains.contains_re(column.obj, prog))
+        elif self.name == pl_expr.StringFunction.Slice:
+            child, expr_start, expr_length = self.children
+            assert isinstance(expr_start, Literal)
+            assert isinstance(expr_length, Literal)
+
```

Comment:
I don't think [str.slice](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.str.slice.html) in polars actually accepts columns for `offset` and `length`

Reply:
Ah thanks @wence- I had the incantation slightly wrong. I'll make some updates here.

Reply:
Running into a bit of a snag here - it seems that the column overload of `slice_strings` doesn't handle things quite in the same way we want. In polars there's true negative indexing, so for the example, the last string slices from `[-2:]`, giving us back `sd`. However a negative value for `start` using the column overload of `slice_strings` clamps negative values to zero. I think there'd be at least a couple of binary ops needed to wrap start values correctly. 

This need to adjust things is coming up in a few edge cases. @wence- should I be handling this with pylibcudf binary ops in this area of the code or elsewhere (in the IR?)

Reply:
We can do a bunch of whole-column things. During translate of stringfunction.slice you could (if either thing is a not a literal) emit some new IR that does the right binops and then everything will work automagically.

I wonder if we could implement this in libcudf with a different `Indexelator` functor. I haven't looked at the libcudf strings code.

Reply:
ok - for instance, I think we can guarantee positive indices without disturbing already-positive values via 

```
(current_idx + len(st)) % len(st)
```

However this requires computing the length and then two binary ops. My assumption is that we're OK with that, and it's just a question of where we do it. Is that correct?

Reply:
You could get a column of string lengths `s.str.len()` and the apply the negative value(s) using a binary op to get the actual start/stop position values. I'd rather not create a polars-specific libcudf APIs if it can be helped. I think we over accommodated pandas-specific behavior in many of strings functions.

Reply:
> I think we over accommodated pandas-specific behavior in many of strings functions.

This is reasonable.

@brandon-b-miller: let's just try and implement things here in the evaluator. Can you please check what the correct semantics for polars are before going to implement things. For example, I _think_ that I'm allowed to index with `-some_value_greater_than_the_string_length` (which corresponds to indexing with `0`) and the `length` is allowed to be longer than the string (and it gets clamped).

So, for example:

```
df = pl.DataFrame({"a": ["1234"]}).lazy()

df.select(pl.col("a").str.slice(-8, 1)).collect()


shape: (1, 1)
┌─────┐
│ a   │
│ --- │
│ str │
╞═════╡
│     │
└─────┘

df.select(pl.col("a").str.slice(-8, 5)).collect()

shape: (1, 1)
┌─────┐
│ a   │
│ --- │
│ str │
╞═════╡
│ 1   │
└─────┘
```

Reply:
Am I missing something? We start with an `offset, length` pair, and we need to produce a `start, stop` pair. So with the wrapping/clamping rules in libcudf, does this transformation not work:

```
start = offset + len(st) if offset < 0 else offset # use copy_if_else
stop = start + length
```

A negative offset is handled by adding the length of the string we're slicing, then we can immediately compute the stop of the slice by adding the length of the _slice_.

The libcudf column kernel clamps the start to `[0, len(st))` but treats negative `stop` as equivalent to "the end of the string". So we need to process the `stop` column further as `max(stop, 0)` to so that it's positive, and then the libcudf kernel which clamps to `[0, len(st))` works fine.

For the case of scalar inputs I think we need to do more clamping manually, but maybe the above model works.

Reply:
I think the addition of `copy_if_else` allowing the inclusion of a conditional statement, plus the idea of doing a `max` on the stop column might be enough to link things up - I'll try locally and see if I can get things passing.

Reply:
After some fiddling we decided to split off the column work into a separate PR, since the math it uses requires some pieces of pylibcudf that haven't been implemented yet and need their own sets of changes in cuDF python, tests, etc. So we will proceed forward here with the scalar cases.